File: apps/api/src/repositories/store.js
import pg from 'pg';
import dotenv from 'dotenv';

const { Pool } = pg;
dotenv.config();

const memoryStore = {
  cvAnalyses: [],
  quizAttempts: [],
  leads: []
};

function normalizeDatabaseUrl(rawUrl) {
  if (!rawUrl) {
    return null;
  }

  if (rawUrl.startsWith('postgres://')) {
    return rawUrl.replace('postgres://', 'postgresql://');
  }

  if (rawUrl.startsWith('postgresql+psycopg://')) {
    return rawUrl.replace('postgresql+psycopg://', 'postgresql://');
  }

  if (rawUrl.startsWith('postgresql+psycopg2://')) {
    return rawUrl.replace('postgresql+psycopg2://', 'postgresql://');
  }

  return rawUrl;
}

const databaseUrl = normalizeDatabaseUrl(process.env.DATABASE_URL);
const pool = databaseUrl
  ? new Pool({
      connectionString: databaseUrl,
      ssl: process.env.DATABASE_SSL === 'true' ? { rejectUnauthorized: false } : undefined
    })
  : null;

export function isDatabaseEnabled() {
  return Boolean(pool);
}

export async function initDatabase() {
  if (!pool || !['1', 'true', 'yes'].includes(String(process.env.AUTO_CREATE_TABLES ?? 'true').toLowerCase())) {
    return;
  }

  await pool.query(`
    CREATE TABLE IF NOT EXISTS users (
      id SERIAL PRIMARY KEY,
      name VARCHAR(120) NOT NULL,
      email VARCHAR(255) NOT NULL UNIQUE,
      created_at TIMESTAMP NOT NULL DEFAULT NOW()
    );

    CREATE TABLE IF NOT EXISTS cvs (
      id SERIAL PRIMARY KEY,
      user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
      file_url TEXT NOT NULL,
      file_name VARCHAR(255) NOT NULL,
      created_at TIMESTAMP NOT NULL DEFAULT NOW()
    );

    CREATE TABLE IF NOT EXISTS skills (
      id SERIAL PRIMARY KEY,
      name VARCHAR(120) NOT NULL UNIQUE
    );

    CREATE TABLE IF NOT EXISTS user_skills (
      user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
      skill_id INTEGER NOT NULL REFERENCES skills(id) ON DELETE CASCADE,
      proficiency SMALLINT NOT NULL DEFAULT 1,
      source VARCHAR(40) NOT NULL DEFAULT 'cv',
      PRIMARY KEY (user_id, skill_id)
    );

    CREATE TABLE IF NOT EXISTS learning_paths (
      id SERIAL PRIMARY KEY,
      user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
      recommendation JSONB NOT NULL,
      created_at TIMESTAMP NOT NULL DEFAULT NOW()
    );

    CREATE TABLE IF NOT EXISTS quiz_attempts (
      id SERIAL PRIMARY KEY,
      user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
      score SMALLINT NOT NULL,
      result JSONB NOT NULL,
      created_at TIMESTAMP NOT NULL DEFAULT NOW()
    );

    CREATE TABLE IF NOT EXISTS leads (
      id SERIAL PRIMARY KEY,
      email VARCHAR(255) NOT NULL,
      target_role VARCHAR(120),
      created_at TIMESTAMP NOT NULL DEFAULT NOW()
    );
  `);
}

async function getOrCreateDemoUser(client) {
  const email = 'demo@skillmap.local';
  const existing = await client.query('SELECT id FROM users WHERE email = $1', [email]);

  if (existing.rows[0]) {
    return existing.rows[0].id;
  }

  const created = await client.query(
    'INSERT INTO users (name, email) VALUES ($1, $2) RETURNING id',
    ['Demo User', email]
  );

  return created.rows[0].id;
}

async function getOrCreateSkill(client, skillName) {
  const existing = await client.query('SELECT id FROM skills WHERE name = $1', [skillName]);

  if (existing.rows[0]) {
    return existing.rows[0].id;
  }

  const created = await client.query('INSERT INTO skills (name) VALUES ($1) RETURNING id', [skillName]);
  return created.rows[0].id;
}

export async function saveCvAnalysis({ fileName, fileSize = 0, analysis }) {
  const record = {
    id: memoryStore.cvAnalyses.length + 1,
    fileName,
    fileSize,
    analysis,
    createdAt: new Date().toISOString()
  };

  memoryStore.cvAnalyses.push(record);

  if (!pool) {
    return record;
  }

  const client = await pool.connect();
  try {
    await client.query('BEGIN');
    const userId = await getOrCreateDemoUser(client);
    const cvResult = await client.query(
      'INSERT INTO cvs (user_id, file_url, file_name) VALUES ($1, $2, $3) RETURNING id',
      [userId, `/uploads/${fileName}`, fileName]
    );

    for (const skillName of analysis.extractedSkills || []) {
      const skillId = await getOrCreateSkill(client, skillName);
      await client.query(
        `
          INSERT INTO user_skills (user_id, skill_id, proficiency, source)
          VALUES ($1, $2, $3, $4)
          ON CONFLICT (user_id, skill_id)
          DO UPDATE SET proficiency = EXCLUDED.proficiency, source = EXCLUDED.source
        `,
        [userId, skillId, 2, 'cv']
      );
    }

    await client.query('INSERT INTO learning_paths (user_id, recommendation) VALUES ($1, $2)', [
      userId,
      JSON.stringify(analysis)
    ]);
    await client.query('COMMIT');

    return { ...record, userId, cvId: cvResult.rows[0].id };
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
}

export async function saveQuizResult({ score, result }) {
  const record = {
    id: memoryStore.quizAttempts.length + 1,
    score,
    result,
    createdAt: new Date().toISOString()
  };

  memoryStore.quizAttempts.push(record);

  if (!pool) {
    return record;
  }

  const client = await pool.connect();
  try {
    const userId = await getOrCreateDemoUser(client);
    await client.query('INSERT INTO quiz_attempts (user_id, score, result) VALUES ($1, $2, $3)', [
      userId,
      score,
      JSON.stringify(result)
    ]);
    return { ...record, userId };
  } finally {
    client.release();
  }
}

export async function saveLead({ email, targetRole }) {
  const record = {
    id: memoryStore.leads.length + 1,
    email,
    targetRole,
    createdAt: new Date().toISOString()
  };

  memoryStore.leads.push(record);

  if (!pool) {
    return record;
  }

  await pool.query('INSERT INTO leads (email, target_role) VALUES ($1, $2)', [email, targetRole || null]);
  return record;
}

export async function getLatestActivity() {
  if (!pool) {
    return {
      cvAnalyses: memoryStore.cvAnalyses.slice(-5).reverse(),
      quizAttempts: memoryStore.quizAttempts.slice(-5).reverse(),
      leads: memoryStore.leads.slice(-5).reverse()
    };
  }

  const [cvAnalyses, quizAttempts, leads] = await Promise.all([
    pool.query('SELECT id, file_name AS "fileName", created_at AS "createdAt" FROM cvs ORDER BY created_at DESC LIMIT 5'),
    pool.query('SELECT id, score, result, created_at AS "createdAt" FROM quiz_attempts ORDER BY created_at DESC LIMIT 5'),
    pool.query('SELECT id, email, target_role AS "targetRole", created_at AS "createdAt" FROM leads ORDER BY created_at DESC LIMIT 5')
  ]);

  return {
    cvAnalyses: cvAnalyses.rows,
    quizAttempts: quizAttempts.rows,
    leads: leads.rows
  };
}

export async function closeDatabase() {
  if (pool) {
    await pool.end();
  }
}

File: apps/api/src/repositories/store.py
from datetime import datetime
from sqlalchemy import text
from src.db import ENGINE, session_scope, get_or_create_demo_user, get_or_create_skill, Cv, UserSkill, LearningPath, QuizAttempt

memory_store = {
    "cvAnalyses": [],
    "quizAttempts": [],
    "leads": []
}

def is_database_enabled():
    return ENGINE is not None

def save_cv_analysis(file_name, file_size=0, analysis=None):
    if analysis is None:
        analysis = {}

    record = {
        "id": len(memory_store["cvAnalyses"]) + 1,
        "fileName": file_name,
        "fileSize": file_size,
        "analysis": analysis,
        "createdAt": datetime.utcnow().isoformat() + "Z"
    }

    memory_store["cvAnalyses"].append(record)

    if not is_database_enabled():
        return record

    try:
        with session_scope() as session:
            user = get_or_create_demo_user(session)

            cv_record = Cv(user_id=user.id, file_url=f"/uploads/{file_name}", file_name=file_name)
            session.add(cv_record)
            session.flush()

            for skill_name in analysis.get("extractedSkills", []):
                skill = get_or_create_skill(session, skill_name)
                # handle upsert or ignore, simplified here:
                existing_skill = session.query(UserSkill).filter_by(user_id=user.id, skill_id=skill.id).first()
                if existing_skill:
                    existing_skill.proficiency = 2
                    existing_skill.source = "cv"
                else:
                    session.add(UserSkill(user_id=user.id, skill_id=skill.id, proficiency=2, source="cv"))

            session.add(
                LearningPath(
                    user_id=user.id,
                    recommendation=analysis,
                )
            )

            return {**record, "userId": user.id, "cvId": cv_record.id}
    except Exception as e:
        print("Database error in save_cv_analysis:", e)
        return record

def save_quiz_result(score, result):
    record = {
        "id": len(memory_store["quizAttempts"]) + 1,
        "score": score,
        "result": result,
        "createdAt": datetime.utcnow().isoformat() + "Z"
    }

    memory_store["quizAttempts"].append(record)

    if not is_database_enabled():
        return record

    try:
        with session_scope() as session:
            user = get_or_create_demo_user(session)
            session.add(QuizAttempt(user_id=user.id, score=score, result=result))
            return {**record, "userId": user.id}
    except Exception as e:
        print("Database error in save_quiz_result:", e)
        return record

def save_lead(email, target_role=None):
    record = {
        "id": len(memory_store["leads"]) + 1,
        "email": email,
        "targetRole": target_role,
        "createdAt": datetime.utcnow().isoformat() + "Z"
    }

    memory_store["leads"].append(record)

    if not is_database_enabled():
        return record

    try:
        with ENGINE.connect() as conn:
            conn.execute(
                text("INSERT INTO leads (email, target_role) VALUES (:email, :target_role)"),
                {"email": email, "target_role": target_role}
            )
            conn.commit()
    except Exception as e:
        print("Database error in save_lead:", e)
    
    return record

def get_latest_activity():
    if not is_database_enabled():
        return {
            "cvAnalyses": list(reversed(memory_store["cvAnalyses"][-5:])),
            "quizAttempts": list(reversed(memory_store["quizAttempts"][-5:])),
            "leads": list(reversed(memory_store["leads"][-5:]))
        }

    try:
        with ENGINE.connect() as conn:
            cvs = conn.execute(text('SELECT id, file_name AS "fileName", created_at AS "createdAt" FROM cvs ORDER BY created_at DESC LIMIT 5')).mappings().all()
            quizzes = conn.execute(text('SELECT id, score, result, created_at AS "createdAt" FROM quiz_attempts ORDER BY created_at DESC LIMIT 5')).mappings().all()
            try:
                leads = conn.execute(text('SELECT id, email, target_role AS "targetRole", created_at AS "createdAt" FROM leads ORDER BY created_at DESC LIMIT 5')).mappings().all()
            except Exception:
                leads = []

            # Format datetime objects for JSON serialization
            def format_rows(rows):
                res = []
                for row in rows:
                    r = dict(row)
                    if "createdAt" in r and r["createdAt"]:
                        r["createdAt"] = r["createdAt"].isoformat() + "Z"
                    res.append(r)
                return res

            return {
                "cvAnalyses": format_rows(cvs),
                "quizAttempts": format_rows(quizzes),
                "leads": format_rows(leads)
            }
    except Exception as e:
        print("Database error in get_latest_activity:", e)
        return {
            "cvAnalyses": list(reversed(memory_store["cvAnalyses"][-5:])),
            "quizAttempts": list(reversed(memory_store["quizAttempts"][-5:])),
            "leads": list(reversed(memory_store["leads"][-5:]))
        }
